File: LeetCodeSolutions/LeetCode_0015.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        n = len(nums)
        if n < 3:
            return []
        res = []
        for left in range(n - 2):
            middle, right = left + 1, n - 1
            if nums[left] > 0 or nums[right] < 0:
                break
            if left > 0 and nums[left - 1] == nums[left]:
                continue
            while middle < right:
                curr_sum = nums[left] + nums[middle] + nums[right]
                if curr_sum > 0:
                    while right - 1 > 0 and nums[right] == nums[right - 1]:
                        right -= 1
                    right -= 1
                elif curr_sum < 0:
                    while middle + 1 < right and nums[middle] == nums[middle + 1]:
                        middle += 1
                    middle += 1
                else:
                    res.append([nums[left], nums[middle], nums[right]])
                    while middle + 1 < right and nums[middle] == nums[middle + 1]:
                        middle += 1
                    middle += 1
                    while right - 1 > 0 and nums[right] == nums[right - 1]:
                        right -= 1
                    right -= 1
        return res

File: LeetCodeSolutions/LeetCode_0015.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int> > ret;
        if (nums.size() < 3) return {};
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for(int l = 0; l < n - 2; ++l){
            if (nums[l] > 0) break;
            if (l > 0 && nums[l] == nums[l - 1]) continue;
            int m = l + 1, r = n - 1;
            while (m < r){
                int s = nums[l] + nums[m] + nums[r];
                if (s < 0){
                    while (m < r && nums[m] == nums[m + 1]) ++m;
                    ++m;
                }
                else if(s > 0){
                    while (r > m && nums[r] == nums[r - 1]) --r;
                    --r;
                }
                else{
                    ret.push_back(vector<int>{nums[l], nums[m], nums[r]});
                    while (m < r && nums[m] == nums[m + 1]) ++m;
                    ++m;
                    while (r > m && nums[r] == nums[r - 1]) --r;
                    --r;
                }
            }
        }
        return ret;
    }
};